### DISCUSSION TITLE
getStaticProps did't return data

### DISCUSSION THREAD

**Discussion by richforest0506** [vercel/next.js#28296] [Help]
I am trying to fetch data from Storyblok API, but in getStaticProps I get the data from the storybook but it can't return data to the page.

pages/docs/[slug].js

```
  import`` React from 'react'
  import Storyblok from "../../lib/storyblok"
   
  export default function DocsArticle( props ) {  
    console.log("PROPS: ", props)
    return (
      <>
          <div className="page">
            {props.story.title}
          </div>
      </>
    );
  }
   
  export async function getStaticProps({ params, preview = false }) {
   
    let slug = params.slug ? params.slug : "home";
    let sbParams = {
      version: "draft", // or 'published' / ' draft
    };
   
    if (preview) {
      sbParams.version = "draft";
      sbParams.cv = Date.now();
    }
  
    // load the stories insides the pages folder
    let { data } = await Storyblok.get(`cdn/stories/${slug}`, sbParams);
    console.log("STORY DATA:", data);
    return {
      props: {
        story: data ? data.story : null,
        preview,
      },
      revalidate: 10, // revalidate 10 seconds
    };
  }
   
  export async function getStaticPaths() {
    
    let { data } = await Storyblok.get('cdn/links/', {
        starts_with: 'article'
    })
    let paths = [];
    Object.keys(data.links).forEach((linkKey) => {
      // don't create routes for folders and the index page
      if (data.links[linkKey].is_folder || data.links[linkKey].slug === "home") {
        return;
      }

      const slug = data.links[linkKey].slug;
          
      let newSlug = slug.replace('docs', '')
      let splittedSlug = newSlug.split("/");
  
      paths.push({ params: { slug: splittedSlug.toString() } });
    });
    
    return {
      paths: paths,
      fallback: false,
    };
  }
}
```

If I display the data using console.log(), I get the data, but it doesn't return it.

```
  STORY DATA: {
    story: {
      name: 'article1',
      created_at: '2021-08-13T00:36:04.648Z',
      published_at: '2021-08-15T16:24:54.810Z',
      id: 66059334,
      uuid: '900a311a-2ad4-461c-9304-e2f36fd25b07',
      content: {
        _uid: 'd21214b5-1e80-4c6a-aa74-259f082a8242',
        content: [Array],
        component: 'page',
        _editable: '<!--#storyblok#{"name": "page", "space": "122749", "uid": "d21214b5-1e80-4c6a-aa74-259f082a8242", "id": "66059334"}-->'
      },
      slug: 'article1',
      full_slug: 'article1',
      sort_by_date: null,
      position: 10,
      tag_list: [],
      is_startpage: false,
      parent_id: 0,
      meta_data: null,
      group_id: 'bdf123cc-0044-4d02-b4b3-28034ee457d0',
      first_published_at: '2021-08-14T00:02:05.000Z',
      release_id: null,
      lang: 'default',
      path: 'docs',
      alternates: [],
      default_full_slug: null,
      translated_slugs: null
    },
    cv: 1629044699,
    rels: [],
    links: []
  }
  PROPS:  {}
  TypeError: Cannot read property 'title' of undefined
```

I'd appreciate it if you could let me know what I did wrong.

**Comment by leerob** [comment]
Hey! I would recommend including a reproducible example of this with a GitHub link. At first glance, nothing looks off, but we'd likely need to see the full code.

**Reply by amochuko** [reply]
@leerob Greetings. I had similar issue while working through the getting started on https://nextjs.org/learn/basics/data-fetching/implement-getstaticprops.

Here's the code in brief from: index.js:

import { getSortedPostsData } from '../lib/posts';

export async function getStaticProps() {

  const allPostsData = getSortedPostsData();

 // NB: This is the returned result when logged to terminal
  //
  // allData: [
  //   {
  //     id: 'ssg-ssr',
  //     title: 'When to Use Static Generation v.s. Server-side Rendering',
  //     date: '2020-01-02',
  //   },
  //   {
  //     id: 'pre-rendering',
  //     title: 'Two Forms of Pre-rendering',
  //     date: '2020-01-01',
  //   },
  // ];

  return {
    props: { allPostsData },
  };
}

export default function Home({allPostsData}) {
  console.log('allData:', allPostsData);

  // NB: This Home component doesn't receive the props as 
  // expected and is 'undefined' when logged
}

**Reply by leerob** [reply]
I see an `async` function but no `await` - do you need to have `await getSortedPostsData()`?

**Reply by leerob** [reply]
A full code reproduction is needed to look further 👍

**Reply by amochuko** [reply]
> I see an `async` function but no `await` - do you need to have `await getSortedPostsData()`?


Thanks got the reply.
Yes, you did see but the resulting function (getSortedPostsData()) isn't async so there was no need to use await.

Below is the code structure:

1. posts/pre-rendering.md
 ---
title: 'Two Forms of Pre-rendering'
date: '2020-01-01'
---

Next.js has two forms of pre-rendering: **Static Generation** and **Server-side Rendering**. The difference is in **when** it generates the HTML for a page.

- **Static Generation** is the pre-rendering method that generates the HTML at **build time**. The pre-rendered HTML is then _reused_ on each request.
- **Server-side Rendering** is the pre-rendering method that generates the HTML on **each request**.

Importantly, Next.js lets you **choose** which pre-rendering form to use for each page. You can create a "hybrid" Next.js app by using Static Generation for most pages and using Server-side Rendering for others.
 

2. pages/posts.js

import fs from 'fs';
import matter from 'gray-matter';
import path from 'path';

const postsDirectory = path.join(process.cwd(), 'posts');

export function getSortedPostsData() {

  const fileNames = fs.readdirSync(postsDirectory);
  const allPostsData = fileNames.map((fileName) => {
    const id = fileName.replace(/\.md$/, '');

    // read markdown file as string
    const fullPath = path.join(postsDirectory, fileName);
    const fileContents = fs.readFileSync(fullPath, 'utf-8');

    const res = matter(fileContents);

    return {
      id,
      ...res.data,
    };
  });

  return allPostsData.sort(({ date: a }, { date: b }) => {
    if (a < b) {
      return 1;
    } else if (a > b) {
      return -1;
    } else {
      return 0;
    }
  });
}

2. pages/index.js

import Head from 'next/head';
import Layout, { name, siteTitle } from '../components/layout';
import { getSortedPostsData } from '../lib/posts';
import utilStyles from '../styles/utils.module.css';

export async function getStaticProps() {
  const allPostsData = await getSortedPostsData();
  return {
    props: { allPostsData },
  };
}

export default function Home({ allPostsData }) {

  console.log('allData:', allPostsData); // the 'Destructured' prop is undefined

  return (
    <div className='container'>
      <Layout home>
        <Head>
          <title>{siteTitle}</title>
        </Head>

        <section className={utilStyles.headingMd}>
          <p>{name}</p>
          <p>
            ( This is a sample website - you'll be buidling a site like this on
            <a href='https://nextjs.org/learn'> our Next.js tutorial</a>.)
          </p>
        </section>
        <section className={`${utilStyles.headingMd} ${utilStyles.padding1px}`}>
          <ul className={utilStyles.list}>
            {allPostsData &&
              allPostsData.map(({ id, date, title }) => {
                <li className={utilStyles.listItem} key={id}>
                  {title} <br /> {id} <br /> {date}
                </li>;
              })}
          </ul>
        </section>
      </Layout>

    </div>
  );
}
